xpdacq/xpdacq.py

```diff
@@ -1,823 +1,416 @@
+
+def _summarize(plan):
+    "based on bluesky.utils.print_summary"
+    output = []
+    read_cache = []
+    for msg in plan:
+        cmd = msg.command
+        if cmd == 'open_run':
+            output.append('{:=^80}'.format(' Open Run '))
+        elif cmd == 'close_run':
+            output.append('{:=^80}'.format(' Close Run '))
+        elif cmd == 'set':
+            output.append('{motor.name} -> {args[0]}'.format(motor=msg.obj,
+                                                             args=msg.args))
+        elif cmd == 'create':
+            pass
+        elif cmd == 'read':
+            read_cache.append(msg.obj.name)
+        elif cmd == 'save':
+            output.append('  Read {}'.format(read_cache))
+            read_cache = []
+    return '\n'.join(output)
+
+
+def _update_dark_dict_list(name, doc):
+    """ generate dark frame reference
+
+    This function should be subscribed to 'stop' documents from dark
+    frame runs.
+    """
+    # always grab from glbl state 
+    dark_dict_list = list(glbl._dark_dict_list)
+    # obtain light count time that is already set to glbl.pe1c
+    acq_time = glbl.area_det.cam.acquire_time.get()
+    num_frame = glbl.area_det.images_per_set.get()
+    light_cnt_time = acq_time * num_frame
+
+    dark_dict = {}
+    dark_dict['acq_time'] = acq_time
+    dark_dict['exposure'] = light_cnt_time
+    dark_dict['timestamp'] = doc['time']
+    dark_dict['uid'] = doc['run_start']
+    dark_dict_list.append(dark_dict)
+    glbl._dark_dict_list = dark_dict_list # update glbl._dark_dict_list
+
+
+def take_dark():
+    "a plan for taking a single dark frame"
+    print('INFO: closing shutter...')
+    yield from bp.abs_set(glbl.shutter, 0)
+    if glbl.shutter_control:
+        yield from bp.sleep(2)
+    print('INFO: taking dark frame....')
+    # upto this stage, glbl.pe1c has been configured to so exposure time is
+    # correct
+    acq_time = glbl.area_det.cam.acquire_time.get()
+    num_frame = glbl.area_det.images_per_set.get()
+    computed_exposure = acq_time*num_frame
+    # update md
+    _md = {'sp_time_per_frame': acq_time,
+           'sp_num_frames': num_frame,
+           'sp_computed_exposure': computed_exposure,
+           'sp_type': 'ct',
+           #'sp_uid': str(uuid.uuid4()), # dark plan doesn't need uid
+           'sp_plan_name': 'dark_{}'.format(computed_exposure),
+           'dark_frame':True}
+    c = bp.count([glbl.area_det], md=_md)
+    yield from bp.subs_wrapper(c, {'stop': [_update_dark_dict_list]})
+    print('opening shutter...')
+    yield from bp.abs_set(glbl.shutter, 1)
+    if glbl.shutter_control:
+        yield from bp.sleep(2)
+
+
+def periodic_dark(plan):
+    """
+    a plan wrapper that takes a plan and inserts `take_dark`
+
+    The `take_dark` plan is inserted on the fly before the beginning of
+    any new run after a period of time defined by `glbl.dk_window` has passed.
+    """
+    need_dark = True
+
+    def insert_take_dark(msg):
```

Comment:
Yes, this is the default. The so-called "dark_strategy" argument in `CustomizedRunEngine.__call__` makes this configurable at *execution time*, separate from the rest of the logic of the plan. That's a little different that bluesky's RunEngine, but it makes sense for your workflow, and it matches the way the original xpdAcq injected this option.